# Predicting Fraudulent Credit Card Transaction
#### Paolo G. Hilado

#### <font color='steelblue'>SITUATIONER:</font> Company is interested to implement a machine learning model that will identify fraudulent transactions. The data set provided contains 30 features and 284,807 records. The V-Columns represent components from Principal Component Analysis, the variable labeled Amount unchanged, the variable Class is although numerical (ready for modelling) has 1 which refers to fraudulent transaction whereas 0 is non-fraudulent. 

In [46]:
# import needed libraries
import pandas as pd
import numpy as np
import researchpy as rp
from sklearn.preprocessing import StandardScaler

In [47]:
# load needed data from the CSV file
df = pd.read_csv("creditcard.csv")
df = df.iloc[:,1:]
df.head()

V1        V2        V3        V4        V5        V6        V7  \
0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9       V10  ...       V21       V22       V23       V24  \
0  0.098698  0.363787  0.090794  ... -0.018307  0.277838 -0.110474  0.066928   
1  0.085102 -0.255425 -0.166974  ... -0.225775 -0.638672  0.101288 -0.339846   
2  0.247676 -1.514654  0.207643  ...  0.247998  0.771679  0.909412 -0.689281   
3  0.377436 -1.387024 -0.054952  ... -0.108300  0.005274 -0.190321 -1.175575   
4 -0.270533  0.817739  0.753074  ... -0.009431  0.798278 -0.137458  0.141267   

        V25       V26       V27       V28  Amount  Class  
0  0.128539 -0.189115  0.133558 -0.021053  149.62      0  
1  0.167170  0.125895 -0.008983  0.014724    2.69      0  
2 -0.327642 -0.139097 -0.055353 -0.059752  378.66      0  
3  0.647376 -0.221929  0.062723  0.061458  123.50      0  
4 -0.206010  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 30 columns]

In [170]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 30 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   V1      284807 non-null  float64
 1   V2      284807 non-null  float64
 2   V3      284807 non-null  float64
 3   V4      284807 non-null  float64
 4   V5      284807 non-null  float64
 5   V6      284807 non-null  float64
 6   V7      284807 non-null  float64
 7   V8      284807 non-null  float64
 8   V9      284807 non-null  float64
 9   V10     284807 non-null  float64
 10  V11     284807 non-null  float64
 11  V12     284807 non-null  float64
 12  V13     284807 non-null  float64
 13  V14     284807 non-null  float64
 14  V15     284807 non-null  float64
 15  V16     284807 non-null  float64
 16  V17     284807 non-null  float64
 17  V18     284807 non-null  float64
 18  V19     284807 non-null  float64
 19  V20     284807 non-null  float64
 20  V21     284807 non-null  float64
 21  V22     28

In [48]:
# Standardize all continuous variables
ndf = pd.DataFrame(StandardScaler().fit_transform(df.iloc[:, :29]))
ndf.rename(columns = {28:"Amount"}, inplace = True)
ndf.head()

0         1         2         3         4         5         6  \
0 -0.694242 -0.044075  1.672773  0.973366 -0.245117  0.347068  0.193679   
1  0.608496  0.161176  0.109797  0.316523  0.043483 -0.061820 -0.063700   
2 -0.693500 -0.811578  1.169468  0.268231 -0.364572  1.351454  0.639776   
3 -0.493325 -0.112169  1.182516 -0.609727 -0.007469  0.936150  0.192071   
4 -0.591330  0.531541  1.021412  0.284655 -0.295015  0.071999  0.479302   

          7         8         9  ...        19        20        21        22  \
0  0.082637  0.331128  0.083386  ...  0.326118 -0.024923  0.382854 -0.176911   
1  0.071253 -0.232494 -0.153350  ... -0.089611 -0.307377 -0.880077  0.162201   
2  0.207373 -1.378675  0.190700  ...  0.680975  0.337632  1.063358  1.456320   
3  0.316018 -1.262503 -0.050468  ... -0.269855 -0.147443  0.007267 -0.304777   
4 -0.226510  0.744326  0.691625  ...  0.529939 -0.012839  1.100011 -0.220123   

         23        24        25        26        27    Amount  
0  0.110507  0.246585 -0.392170  0.330892 -0.063781  0.244964  
1 -0.561131  0.320694  0.261069 -0.022256  0.044608 -0.342475  
2 -1.138092 -0.628537 -0.288447 -0.137137 -0.181021  1.160686  
3 -1.941027  1.241904 -0.460217  0.155396  0.186189  0.140534  
4  0.233250 -0.395202  1.041611  0.543620  0.651816 -0.073403  

[5 rows x 29 columns]

In [49]:
# Column binding the continuous variables with the categorical of 0 (non-fraud) and 1 (fraud) transaction
ndf = pd.concat([ndf, df.iloc[:,29]], axis = 1)
ndf.head(5)

0         1         2         3         4         5         6  \
0 -0.694242 -0.044075  1.672773  0.973366 -0.245117  0.347068  0.193679   
1  0.608496  0.161176  0.109797  0.316523  0.043483 -0.061820 -0.063700   
2 -0.693500 -0.811578  1.169468  0.268231 -0.364572  1.351454  0.639776   
3 -0.493325 -0.112169  1.182516 -0.609727 -0.007469  0.936150  0.192071   
4 -0.591330  0.531541  1.021412  0.284655 -0.295015  0.071999  0.479302   

          7         8         9  ...        20        21        22        23  \
0  0.082637  0.331128  0.083386  ... -0.024923  0.382854 -0.176911  0.110507   
1  0.071253 -0.232494 -0.153350  ... -0.307377 -0.880077  0.162201 -0.561131   
2  0.207373 -1.378675  0.190700  ...  0.337632  1.063358  1.456320 -1.138092   
3  0.316018 -1.262503 -0.050468  ... -0.147443  0.007267 -0.304777 -1.941027   
4 -0.226510  0.744326  0.691625  ... -0.012839  1.100011 -0.220123  0.233250   

         24        25        26        27    Amount  Class  
0  0.246585 -0.392170  0.330892 -0.063781  0.244964      0  
1  0.320694  0.261069 -0.022256  0.044608 -0.342475      0  
2 -0.628537 -0.288447 -0.137137 -0.181021  1.160686      0  
3  1.241904 -0.460217  0.155396  0.186189  0.140534      0  
4 -0.395202  1.041611  0.543620  0.651816 -0.073403      0  

[5 rows x 30 columns]

In [45]:
# check to see continuous variables were standardized
rp.summary_cont(ndf.iloc[:, :29])

Variable         N  Mean   SD      SE  95% Conf.  Interval
0       0.0  284807.0  -0.0  1.0  0.0019    -0.0037    0.0037
1       1.0  284807.0   0.0  1.0  0.0019    -0.0037    0.0037
2       2.0  284807.0  -0.0  1.0  0.0019    -0.0037    0.0037
3       3.0  284807.0  -0.0  1.0  0.0019    -0.0037    0.0037
4       4.0  284807.0  -0.0  1.0  0.0019    -0.0037    0.0037
5       5.0  284807.0   0.0  1.0  0.0019    -0.0037    0.0037
6       6.0  284807.0  -0.0  1.0  0.0019    -0.0037    0.0037
7       7.0  284807.0  -0.0  1.0  0.0019    -0.0037    0.0037
8       8.0  284807.0   0.0  1.0  0.0019    -0.0037    0.0037
9       9.0  284807.0   0.0  1.0  0.0019    -0.0037    0.0037
10     10.0  284807.0   0.0  1.0  0.0019    -0.0037    0.0037
11     11.0  284807.0   0.0  1.0  0.0019    -0.0037    0.0037
12     12.0  284807.0  -0.0  1.0  0.0019    -0.0037    0.0037
13     13.0  284807.0  -0.0  1.0  0.0019    -0.0037    0.0037
14     14.0  284807.0  -0.0  1.0  0.0019    -0.0037    0.0037
15     15.0  284807.0  -0.0  1.0  0.0019    -0.0037    0.0037
16     16.0  284807.0   0.0  1.0  0.0019    -0.0037    0.0037
17     17.0  284807.0  -0.0  1.0  0.0019    -0.0037    0.0037
18     18.0  284807.0  -0.0  1.0  0.0019    -0.0037    0.0037
19     19.0  284807.0   0.0  1.0  0.0019    -0.0037    0.0037
20     20.0  284807.0   0.0  1.0  0.0019    -0.0037    0.0037
21     21.0  284807.0   0.0  1.0  0.0019    -0.0037    0.0037
22     22.0  284807.0  -0.0  1.0  0.0019    -0.0037    0.0037
23     23.0  284807.0   0.0  1.0  0.0019    -0.0037    0.0037
24     24.0  284807.0  -0.0  1.0  0.0019    -0.0037    0.0037
25     25.0  284807.0   0.0  1.0  0.0019    -0.0037    0.0037
26     26.0  284807.0  -0.0  1.0  0.0019    -0.0037    0.0037
27     27.0  284807.0   0.0  1.0  0.0019    -0.0037    0.0037
28   Amount  284807.0   0.0  1.0  0.0019    -0.0037    0.0037

In [70]:
# let us now split our data into training and test set
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(ndf.iloc[:, :29], ndf.iloc[:, 29], random_state = 100, test_size  =0.3)
train_ndf = pd.concat([X_train, y_train], axis = 1)
test_ndf = pd.concat([X_test, y_test], axis = 1)

In [88]:
# Check with Logistic Regression
from sklearn.linear_model import LogisticRegression
logreg_clf = LogisticRegression()
logreg_clf.fit(train_ndf.iloc[:, :29], train_ndf.iloc[:, 29])
res = logreg_clf.predict(X_test)

In [150]:
# Let us check the accuracy
# First create a simple function for this
def checkAcc(y_true, y_pred):
    a = sum(y_true == y_pred)/y_true.shape[0]
    return a

print("Model Accuracy is:",np.round(checkAcc(y_test, res)*100,2), "%")

In [160]:
# Saving the true labels and predictions
a = pd.DataFrame(res)
b = pd.DataFrame(y_test)
a.reset_index(drop=True, inplace=True)
b.reset_index(drop=True, inplace=True)
results = pd.concat([b,a], axis =1)
results.rename(columns = {0: "true labels", "Class":"prediction"}, inplace = True)
results.to_csv("CCfraud.csv")

In [163]:
# Save the Model for Future use
import pickle
filename = "CCfraudDet.sav"
pickle.dump(logreg_clf, open(filename, "wb"))

In [168]:
# Load the model
loaded_model = pickle.load(open("CCfraudDet.sav", 'rb'))
result = loaded_model.score(X_test, y_test)
print("Loaded Model Accuracy is:", np.round(result*100,2), "%")

Loaded Model Accuracy is: 99.91 %


#### <font color='steelblue'>RESULT:</font> We now have created an ML model that will allow us to classify whether future transactions would be flagged as a possible fraud. The said model, as shown in the tests conducted, performs with 99.91% accuracy.